### PR TITLE
[JENKINS-69534] Fix a race condition that causes file descriptor leaks when cloud agents are created

### PR DESCRIPTION
**Pull request by jonaslind**
See [JENKINS-69534](https://issues.jenkins.io/browse/JENKINS-69534).

The Jenkins.computers map is changed from a CopyOnWriteMap to a ConcurrentHashMap to simplify concurrent creation of computers. This allows us to make AbstractCIBase's createNewComputerForNode and updateNewComputer methods thread safe and simplify removeComputer.

I've executed some local manual verification but no autotests have been written due to the complexity of testing concurrency.

### Proposed changelog entries

* Fix a race condition that causes file descriptor leaks when cloud agents are created

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change) and are in the imperative mood. [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  - Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadoc, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")` if applicable.
- [x] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content-Security-Policy directives (see [documentation on jenkins.io](https://www.jenkins.io/doc/developer/security/csp/)).
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@basil
@res0nance

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are accurate, human-readable, and in the imperative mood
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7087"></a>

### CODE REVIEW

**Comment by jonaslind** [comment]
One detail I'd like to discuss is predictable iteration order on the computers map.

The previously used CopyOnWriteMap was backed by a LinkedHashMap so iteration order of node-to-computer-mappings was predictable.

The now used ConcurrentHashMap does not guarantee iteration order.

Is there any logic somewhere that relies on the order of node-to-computer-mappings that may get broken by this change?

File: core/src/main/java/hudson/model/AbstractCIBase.java
Comment (res0nance) [inline core/src/main/java/hudson/model/AbstractCIBase.java:208]: This changes the current behaviour where we now remove before locking the queue. We probably should leave this as locking before removing the entry.

This would result in code such as https://github.com/jenkinsci/jenkins/blob/c92096b7f5e735c924c5b5758f9166ade3d3a546/core/src/main/java/hudson/model/Queue.java#L1490 having a Computer that has been removed.

File: core/src/main/java/hudson/model/AbstractCIBase.java
Comment (basil) [inline core/src/main/java/hudson/model/AbstractCIBase.java:163]: This change is good. But the type of the map is still `Map` rather than `ConcurrentMap` in the calls to `#getComputerMap` in `AbstractCIBase#getComputer` and `AbstractCIBase#updateComputerList` further below, which is inconsistent. Can you update those to `ConcurrentMap` for consistency?

**Review by basil — COMMENTED** [review]
Thanks for the PR! This looks good overall. I am still reviewing it in detail, but I wanted to leave some quick feedback after my first pass to get you started. I will do a more thorough review later today.

> Is there any logic somewhere that relies on the order of node-to-computer-mappings that may get broken by this change?

I do not believe so. The iteration over the entries in `AbstractCIBase#removeComputer` is looking for a single unique item, so the order of iteration does not matter. The iteration over the entries in `AbstractCIBase#updateComputerList` processes the entries into a `HashSet` and `HashMap` before proceeding further, so the existing code already works with an undefined order. I could not find any other references to iteration besides these.

**Comment by jonaslind** [comment]
Thanks @res0nance and @basil for feedback! I've updated the commit according to your comments.

**Review by res0nance — APPROVED** [review]
LGTM!

File: core/src/main/java/hudson/model/AbstractCIBase.java
Comment (basil) [inline core/src/main/java/hudson/model/AbstractCIBase.java:179]: Optional non-blocking feedback for your consideration: In the scenario described in [JENKINS-69534](https://issues.jenkins.io/browse/JENKINS-69534), where in thread 1 a `ComputerA1` for `NodeA` fails to launch and `KubernetesLauncher#launch(SlaveComputer, TaskListener)` invokes `AbstractCloudSlave#terminate`, and in thread 2 a timer thread invokes `NodeProvisioner#suggestReviewNow`, resulting in both threads entering `AbstractCIBase#createNewComputerForNode(Node, boolean)` for `NodeB`, the losing thread will log this statement at warning level. Since this behavior is expected, it seems to me that warning level might be too high a level for this log statement, being better reserved for unexpected but non-fatal scenarios. Perhaps a debug or fine log would be more appropriate if this is the expected behavior. On the other hand, if you think this is exceptional enough to warrant a warning, then feel free to leave this code as-is.

**Review by basil — APPROVED** [review]
Nice job! The code looks good to me, pending one optional non-blocking piece of feedback about the log level of the new log statement.

Regarding backward compatibility, I searched for usages of `#getComputerMap` in the whole `jenkinsci` GitHub organization, and I found no references besides Jenkins core itself. So it should not cause any compatibility problems to change the type of the return value from `Map` to `ConcurrentMap`. :+1: 

In [JENKINS-69534](https://issues.jenkins.io/browse/JENKINS-69534), you mentioned that you have locally forked Jenkins in order to verify your original proposal. Are you willing to run the code from this PR in your environment for a short period of time to verify that the issue is resolved and there are no regressions?

File: core/src/main/java/hudson/model/AbstractCIBase.java
Comment (jonaslind) [reply core/src/main/java/hudson/model/AbstractCIBase.java:179]: Thanks!
This log line was taken directly from line `219` of the previous version of `AbstractCIBase` and I didn't reflect on the level when I moved it. I agree that `WARNING` is too high, I'll change it to `FINE`.

**Comment by jonaslind** [comment]
@basil writes:

> In [JENKINS-69534](https://issues.jenkins.io/browse/JENKINS-69534), you mentioned that you have locally forked Jenkins in order to verify your original proposal. Are you willing to run the code from this PR in your environment for a short period of time to verify that the issue is resolved and there are no regressions?

Back when I was initially investigating this I set up a local bubblegum-and-shoestring test environment where I could almost reliably reproduce the issue. I've tested this change in that environment without being able to reproduce the issue.

I cannot test this exact change in our production CI environment because we haven't moved to Java 11 there yet (will do soon!) but I've tested the same change on top of 2.332.2 and it has run for ~8 hours today during our peak loads with stable file descriptors. It would not have been that stable during peak loads if there was a file descriptor leak. I haven't noticed any other issues in our production CI environment either.

**Comment by basil** [comment]
Thank you for contributing this fix and for testing it in your local environment!

---

This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Comment by jonaslind** [comment]
Thanks for all the guidance and input!

**Pull request merged**
